[Kitsune] Added kit-plot to Kitsune utilities

jsarrao opened the pull request:
Kit-plot is a utility script that generates plots from a Kitsune test suite report, based on the options provided by the user. There are 3 main options (and several, less significant options) that a user can provide: benchmarks, metrics, and targets. Benchmarks are the benchmarks the test suite runs (vecadd, srad, euler3d, etc.), metrics are the types of data collected by the test suite (compile time, runtime, etc.), and targets refers to the combination of tapir targets and "external" compilers the test suite tests. For each metric, kit-plot will generate a grouped bar chart showing a comparison of the different targets across all benchmarks.

jsarrao committed:
added kit-plot to test suite utils

jsarrao commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,409 @@
+
+                if metric_name in target:
+                    val = target[metric_name]
+                elif metric_name == "total-runtime":
+                    val = target["times"]["total"]["total"]
+                elif metric_name == "mean-runtime":
+                    val = target["times"]["total"]["mean"]
```
Reading the data like this will fail if a given benchmark is not available for a specific target, but the target does still have data in the report file for other benchmarks (for example if someone is running in benchmark mode, and the opencilk test is taking a very long time so they kill the process in htop). 

Do you think this is worth accounting for? I could do a try/except for when this happens, but given how many different places the data is read, it just feels like it would be a lot of extra code and another level of indentation

tarunprabhu replied:
How about writing a sanity check function that checks for all these things before you start trying to generate anything. If the check fails, you could tell the user exactly why.

Another option would be to use the `.get()` method and give it a "reasonable" default to use in case some data is missing. If combined with a sanity check, you could downgrade the check failure to a warning and say that you will be using defaults when certain data is not available

jsarrao replied:
I looked at the report a bit more closely and there is a reasonable solution here. When a test fails for a target, this is recorded in the report. I previously thought that if it failed, it would just not put any info in the report entry, so the only way to tell if it didn't have data was to handle the resulting `KeyError`. However, since the report records that it failed, I can just check for those that failed and use a default value of 0 in those cases instead.

tarunprabhu replied:
I think I added the pass/fail status to the report because you mentioned something about failures once before, so it may not have been there when you first looked at the report. Sorry, I should have told you about it earlier.

tarunprabhu replied:
It would still be a good idea to warn the user if some data was missing and a default of 0 would to be assumed. It helps avoid some head-scratching if bars are missing from the result.

jsarrao committed:
[kit-plot] modified description for help message

jsarrao committed:
[kit-plot] adjusted y label for baseline plots

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+# Parse the command line arguments. Returns the parsed object if parsing was
+# successful and will terminate the program with an error message otherwise
+def parse_command_line_args():
+    prog = "plotter"
```
Is there a reason that you are overwriting `prog` here?

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+        metavar="BENCHMARK",
+        nargs="+",
+        type=str,
+        choices=["copy", "euler3d", "raytracer", "saxpy", "srad", "vecadd"],
```
It might be good to make this list a global variable. That way, if we add a new benchmark, we can easily find the global at the top of the file and add to it.

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+        type=str,
+        metavar="METRIC",
+        nargs="+",
+        choices=["compile-time", "link-time", "total-runtime", "mean-runtime"],
```
This should be made global as well. Since these are constants, it is reasonable to do so.

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+    parser.add_argument(
+        "--no-annotate",
+        action="store_false",
+        help="Enable y-value annotations above data bars",
```
Should this be disable?

```suggestion
        help="Disable y-value annotations above data bars",
```

jsarrao replied:
I first had the annotations as turned off by default, but then when I made them default for the baseline graphs I forgot to change the help message...

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+        choices=["copy", "euler3d", "raytracer", "saxpy", "srad", "vecadd"],
+        default=["copy", "euler3d", "raytracer", "saxpy", "srad", "vecadd"],
+        help="List of benchmarks to include in plots. "
+        "If bechmarks are not specified, data from all possible "
```
What is your metric for determining when to add newlines? Is it the width of your terminal? Is it assuming a terminal with a width of 80 characters?

jsarrao replied:
In general, yes I use a width of 80 characters. I think for specific example, I had written more on the first line but decided to delete it. I kind of thought the formatter I ran would take care of stuff like that so I didn't look super closely but I guess I assumed incorrectly...

tarunprabhu replied:
I would keep this simple and let `argparse` handle everything. If this were a user-facing tool, I would be more inclined to tweak the output, but since it is not, it's probably better to keep things "vanilla"

tarunprabhu replied:
What I mean is to stop using explicit newlines in the help messages and don't provide a custom formatter.

jsarrao replied:
Oh I thought you were refering to how I split the string on lines 109 and 110. But yes, since argparse is now actually displaying the options automatically, I don't need to list the options, so I don't need the newlines, and I can take the custom formatter out which resolves that issue

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+        choices=["copy", "euler3d", "raytracer", "saxpy", "srad", "vecadd"],
+        default=["copy", "euler3d", "raytracer", "saxpy", "srad", "vecadd"],
+        help="List of benchmarks to include in plots. "
+        "If bechmarks are not specified, data from all possible "
```
```suggestion
        "If benchmarks are not specified, data from all possible "
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+        help="Path to directory where you would like to save plots (default:"
+        " save to current directory)",
+    )
+    # TODO: parsing the arguments into a list like this is very clean, but it
```
If you add some text to the help message (either in the description or the epilog of the argument parser object about using `--` in the command, this comment will become redundant.

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+import os
+import sys
+
+# Despite representing data for different metrics, all the plots are the same
```
I would update the comment to include what each element in the tuple represents. I think this "Title, Y-label, X-label, and filename"

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+# metric name as the key
+plot_labels = {
+    "compile-time": (
+        "Compilation Time Comparison",
```
Perhaps "Total Compile Time" is better as a title

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+    "compile-time": (
+        "Compilation Time Comparison",
+        "Time (seconds)",
+        "Benchmark Name",
```
I think just "Benchmark" instead of "Benchmark Name" is sufficient. There is a constant battle between brevity and descriptiveness. There are never any winners (sigh)

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+        "compile-time.png",
+    ),
+    "link-time": (
+        "Link Time Comparison",
```
In general, the term "Comparison" is unnecessary. Just "Link Time" should be fine. The same for the subsequent titles.

jsarrao replied:
My motivation here is that the example plots that Pat sent me used "Comparison" in the title and "Benchmark Name" in the x axis label, but I am happy to change these.

tarunprabhu replied:
I see. Well, leave them as they are then. Since Pat is the one who will usually be presenting these, let's stick with what he started out doing.

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+        "mean-runtime.png",
+    ),
+    "euler3d-total-runtime": (
+        "Euler3d Kernel Performance Comparison: Total Runtime",
```
Does this generate a bar for each kernel in Euler3d? If so, perhaps "Euler3d Total Runtime (Kernels)"

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+        "Euler3d Kernel Performance Comparison: Total Runtime",
+        "Total Runtime (seconds)",
+        "Kernel Name",
+        "euler3d-kernel-total-rt.png",
```
Please be consistent with the filenames. If you have used "runtime" elsewhere, use "runtime" here as well. It's ok if the names are a bit long.

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+# successful and will terminate the program with an error message otherwise
+def parse_command_line_args():
+    prog = "plotter"
+    descr = "Generate plots from Kitsune test suite report. Will overwrite any plot files currently in save directory."
```
```suggestion
    descr = "Generate plots from the Kitsune test suite report. This will overwrite any plot files currently in the save directory."
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+        "-k",
+        "--kernel-plots",
+        action="store_true",
+        help="When used, plots mean kernel runtime plots for benchmarks "
```
```suggestion
        help="When used, plots the mean kernel runtime for benchmarks "
```

I would also suggest making a list of the benchmarks containing multiple kernels and making it global. You can use that list to generate the help string for this option. For instance
```
"with multiple kernels" + ', '.join(benchmarks_multi_kernel)
```
This makes it easier to keep this script in sync if we add new benchmarks to the test suite

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+        choices=["compile-time", "link-time", "total-runtime", "mean-runtime"],
+        default=["compile-time", "link-time", "total-runtime", "mean-runtime"],
+        help="List of metrics to generate plots for. "
+        "If metrics flag is not specified, plots for all possible "
```
```suggestion
        "If this option is not specified, plots for all known "
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+        help="List of metrics to generate plots for. "
+        "If metrics flag is not specified, plots for all possible "
+        "metrics will be generated\n"
+        "options: compile-time, link-time, total-runtime, "
```
Is is necessary to list the options here? I think if an explicit `choices` argument is provided, are they not rendered automatically?

jsarrao replied:
At least when I print out the help message, the `choices` are not automatically displayed. The only way I found to display the choices is if you try to pass something that isn't one of the choices, in which case it will give you an error message for passing an unsupported option as well as printing the list

tarunprabhu replied:
Weird. I just tried a quick test and here is what I got:

```python
#!/usr/bin/env python3

import argparse as ap

parser = ap.ArgumentParser()
parser.add_argument('--list', choices=['a', 'b', 'c'], nargs='+')

args = parser.parse_args()
print(args.list)
```

```
$ ./test.py -h
usage: test.py [-h] [--list {a,b,c} [{a,b,c} ...]]

options:
  -h, --help            show this help message and exit
  --list {a,b,c} [{a,b,c} ...]
```

jsarrao replied:
Ah it's because I am setting the metavar. If I take that out it prints the lists like you see there

tarunprabhu replied:
Ah! I had no idea that it worked that way. There you go! I learned something new today. Thank you.

[edit: Emphasize that *I* learned something new today]

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+        help="List of benchmarks to include in plots. "
+        "If bechmarks are not specified, data from all possible "
+        "benchmarks will be plotted\n"
+        "options: copy, euler3d, raytracer, saxpy, srad, vecadd",
```
Is it necessary to list the options explicitly here despite a `choices` argument being provided

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+        metavar="TARGET",
+        nargs="+",
+        help="List of targets to include in plots. "
+        "If metrics flag is not specified, plots using all possible "
```
```suggestion
        "If this option is not specified, plots using all known "
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+        help="List of targets to include in plots. "
+        "If metrics flag is not specified, plots using all possible "
+        "targets will be generated\n"
+        "options are based on your architecture, use --print-targets "
```
I assume that "targets" here is a combination of tapir targets and "external" compilers, right? For instance, is "nvcc" a valid "target" here? Maybe we should think of an alternative term for this and update the message accordingly. Any suggestions?

jsarrao replied:
"Target" is the wording used in report.json, but yes it refers to the combination of tapir targets and "external" compilers

tarunprabhu replied:
(sigh) Yeah, this is a mess. The term "target" is only marginally less overloaded than "cuda" in our case. 

I would drop the phrase "options are based on your architecture" because that is not correct. The targets in the report are based on how the test suite was configured. It may reflect the "architecture", but it need not. I would just say "Use --print-targets ..."

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+        help="Enable y-value annotations above data bars",
+    )
+    parser.add_argument(
+        "--no-baseline", action="store_true", help="Turn off baseline mode for plots"
```
If `--no-baseline` is provided, will this generate plots with raw runtimes? The help message is not too clear about this.

jsarrao replied:
That is correct, will update

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+    parser.add_argument(
+        "--no-metadata",
+        action="store_false",
+        help="Turns metadata caption with date/time and GPU info off",
```
```suggestion
        help="Do not generate the caption containing metadata such as date, time, GPU model etc.",
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+    parser.add_argument(
+        "--print-targets",
+        action="store_true",
+        help="If this flag is used, no plots will be generated. Instead a "
```
```suggestion
        help="Print the list of targets present in the report. Do not generate any plots"
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+    # Collect baseline data for plots
+    if not args.no_baseline:
+        possible_baselines = ["kokkos-nvidia", "kokkos-amd"]
+        # TODO: I feel like there should be a better way of doing this but if
```
Would a loop with early termination if a baseline was found be "better"? This actually looks reasonable to me. The only way it would be slightly more elegant would be if you moved this code into a function
```
baseline = get_baseline()
```
The function could then have a loop with a return instead of a break.

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+        )
+
+    for index, target_name in enumerate(sorted(data.keys())):
+        # This is somewhat convoluted.
```
Is this still convoluted? :-)

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+
+    for index, target_name in enumerate(sorted(data.keys())):
+        # This is somewhat convoluted.
+        # The benchmarks are not run in alphabetical order. In practice, this
```
I think the order in which the benchmarks are run has no bearing on this, but the order in which they are returned when you look them up in the dictionary. 

Other than that, I like this comment because it explains exactly why you are sorting this.

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+        if args.no_baseline and metric in args.error_bars_data:
+            error_mins = []
+            error_maxs = []
+            # Similar idea to vals, but the resulting list needs to be split
```
This is another example of a good comment. You explain exactly why are doing something that would otherwise leave readers not very familiar with matplotlib scratching their heads

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+            vals,
+            width,
+            label=target_name,
+            zorder=3,
```
Would it make sense to make these (3 for `zorder` and 5 for `capsize`) named constants instead of just using the values here. There are other constants that you have used here that seem to affect the appearance of the final plot. If these could be turned into named global variables, you could have documentation at the top of the file describing how they could be adjusted to affect the final appearance. That way, if we needed a one-off plot that needed to be a bit different, it would be easier to do.

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+
+    # Clean up the layout and save
+    fig.tight_layout()
+    plt.savefig(filename, dpi=300)
```
We may want to make the DPI configurable. I think it may be worth providing a command line option. It may help when generating images for posters, for example. The default of 300 is fine.

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+                elif metric_name == "mean-runtime":
+                    val = target["times"]["total"]["mean"]
+                    if args.error_bars:
+                        rt_min = target["times"]["total"]["min"] / 1000000
```
Maybe 100000 could be made a constant: e.g. `US_PER_SEC = 1000000`. That way, if we need to plot, say, milliseconds, we can just tweak the value instead of changing it in several places. 

In general, if you ever find yourself hard-coding integer/string literals anywhere in the code, stop to ask yourself if it is worth hoisting them up to a named constant. It often is (but not always).

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+    if args.error_bars and not args.no_baseline:
+        print(
+            "Warning: Error bars were enabled but it is not possible to "
+            "include error bars on plots scaled by a baseline. Generated "
```
Why are error bars not possible on plots that have been normalized to a baseline?

jsarrao replied:
Well I guess it wasn't that it wasn't possible, but it made the graphs look really weird since nvcc's max was so much larger than the baselines max.

In any event, I don't think we can do error bars at all any more since we are filtering out the max value and I have no way of knowing what the new max is

tarunprabhu replied:
I think filtering out max should be configurable, and opt-in. The default should not be to drop any such value because we are likely to forget that that is happening and we could get into trouble in the future. In fact, I would actually get rid of both max and min for symmetry in such cases.

jsarrao replied:
Sounds good. I will add the error bar functionality for the mean runtime baseline plots, make the max filtering opt-in, and then drop the error bars only when filtering out the max (and update this warning to reflect that)

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,405 @@
+        )
+
+    if args.error_bars and not args.no_annotate:
+        print(
```
It is good practice to print warning messages to `stderr` instead of `stdout`. That way, if the script is ever tweaked to print something to stdout, stderr can be redirected to `/dev/null` (for instance) to avoid it "polluting" the output.

tarunprabhu reviewed (changes requested):
Please follow the [LLVM conventions](https://llvm.org/docs/DeveloperPolicy.html#commit-messages) for commit messages. 

Please update the title accordingly. A tag of [Kitsune] will be suitable here. Also update the description of this PR with a description of kit-plot, what it does and what it is used for. The title of the PR will become the "title" referenced in the developer policy, the description of the PR will be the commit message.

jsarrao committed:
[kit-plot] Added extrema filtering, adjusted error bar functionality, and improved code style/format

  - Added new command line option that allows users to filter out min and max values
  - Added functionality to allow error bars to be included in baseline plots
  - Now use a default value when a target is missing data for a given benchmark
  - Lifted many constant values to named global constants
  - Added several new error checks and messages
  - Warnings and errors now printed to stderr
  - Fixed grammar and imporved clarity of help message

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -6,15 +6,17 @@ import matplotlib.pyplot as plt
 # metric. This dictionary stores this information for each metric by using the
 # metric name as the key
-plot_labels = {
+# Tuple entries are of form (Title, Y-label, X-Label, Filename)
```
```suggestion
# Tuple entries are of the form (Title, Y-label, X-Label, Filename)
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -86,80 +96,113 @@ def parse_command_line_args():
+        "--print-targets to print options",
+    )
+    parser.add_argument(
+        "--image_dpi",
```
Don't use underscores in long options.

```suggestion
        "--image-dpi",
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -86,80 +96,113 @@ def parse_command_line_args():
+        "--no-baseline",
+        action="store_true",
+        help="Do not process data with a baseline target. Instead, plots will "
+        "use raw data"
```
```suggestion
        help="Do not normalize metrics to a baseline when plotting. Instead, "
        "use absolute values of the metrics"
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -86,80 +96,113 @@ def parse_command_line_args():
+        help="Path to .json report file. Note: if this argument is placed "
+        "immediately after one of the benchmarks, metrics, or targets "
+        "options, a -- must be placed in between (eg. --benchmarks ... "
+        "-- /path/to/report.json)"
```
How about something like this instead?

```suggestion
        "immediately after one of --benchmarks, --metrics, or --targets, "
        "'--' must be added before <file> (eg. --benchmarks euler srad "
        "-- /path/to/report.json)"
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -38,43 +40,51 @@ plot_labels = {
+
+ERROR_BAR_CAP_SIZE = 5
+
+FIGURE_WIDTH = 9
+FIGURE_HEIGHT = 6
```
Could you add a comment describing what the units of `FIGURE_WIDTH` and `FIGURE_HEIGHT` are here? If it's some matplotlib magic, maybe just say that. That way, in the future, if we need to change the size, we have some idea of what to change this to.

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -38,43 +40,51 @@ plot_labels = {
 }
 
+BENCHMARK_OPTIONS = ["copy", "euler3d", "raytracer", "saxpy", "srad", "vecadd"]
+METRIC_OPTIONS = ["compile-time", "link-time", "total-runtime", "mean-runtime"]
```
This name makes it consistent with `MULTI_KERNEL_BENCHMARKS` and `MULTI_KERNEL_METRICS`. You could consider naming these `ALL_BENCHMARKS` and `ALL_METRICS` too. Could you add some instructions here. Perhaps something like this:

```suggestion
# All known benchmarks in kitsune-test-suite. 
# If any benchmarks are added to `Kitsune/Benchmarks`, this list should be updated.
BENCHMARKS = ["copy", "euler3d", "raytracer", "saxpy", "srad", "vecadd"]

# All known metrics collected when building/running the kitsune-test-suite. 
# If any additional metrics are collected, this list should be updated.
METRICS = ["compile-time", "link-time", "total-runtime", "mean-runtime"]
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -168,77 +211,79 @@ def generate_plot(args, data, metric):
+            #       might worry. But because the plots look the way we want them
+            #       to, this worry would be completely unneccesary. That being
+            #       said there could be a different possible runtime warning
+            #       that I am unaware of which will now never be seen
```
I think you are right. It is not a great idea to suppress warnings like this. A quick google search got me this:

https://stackoverflow.com/a/37977222

Does that sound like something we could do instead?

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -168,77 +211,79 @@ def generate_plot(args, data, metric):
+            if not args.no_baseline:
+                # Again, suppress divide by zero warnings in the event of
+                # missing data
+                warnings.filterwarnings("ignore", category=RuntimeWarning)
```
Same comment here regarding warnings

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -269,13 +314,20 @@ def generate_plot(args, data, metric):
+        missing_targets = [target for target in args.targets
+                           if target not in json_data["targets"]]
+        if missing_targets:
+            sys.exit("Error: There is no data in provided report for the following targets: " + ', '.join(missing_targets))
```
```suggestion
            sys.exit("Error: There is no data in the report for the following targets: " + ', '.join(missing_targets))
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -315,27 +365,49 @@ def parse_and_plot_benchmark_data(args, json_data):
-                    val = target[metric_name]
+                if target["code"] == "fail":
+                    print(
+                        f"Warning: no test data for {target_name} {benchmark_name}. Bar will be omited from {metric_name} plot",
```
```suggestion
                        f"Warning: no test data for {target_name} {benchmark_name}. {target_name} will be omited from {metric_name} plot",
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -315,27 +365,49 @@ def parse_and_plot_benchmark_data(args, json_data):
+                    value = 0
+                    if args.error_bars and metric_name == "mean-runtime":
+                        args.error_bars_data[metric_name][target_name][
+                            benchmark_name
```
Yuck! Which formatter did this?

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -315,27 +365,49 @@ def parse_and_plot_benchmark_data(args, json_data):
                     plot_data[metric_name][target_name] = {}
```
No need to change them now, but you could have just used `bname`, `mname`, and `tname` for `benchmark_name`, `metric_name` and `target_name`. In context, it is not too difficult to guess what names they refer to.

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -315,27 +365,49 @@ def parse_and_plot_benchmark_data(args, json_data):
                     continue
+                if target["code"] == "fail":
+                    print(
+                        f"Warning: no kernel data for {target_name} {benchmark_name}. Skipping target for {metric_name} plot",
```
```suggestion
                        f"Warning: no kernel data for {target_name} {benchmark_name}. Skipping {target_name} in {metric_name} plot",
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -86,80 +96,113 @@ def parse_command_line_args():
+        help="Value to be used for dpi of saved images",
+    )
+    parser.add_argument(
+        "--filter_extremes",
```
```suggestion
        "--filter-extremes",
```

tarunprabhu reviewed (changes requested):
Thanks. This is coming together nicely. Most issues are matters of style and typos. We should find another way of dealing with zeros rather than suppressing divide by zero warnings.

[edit] I also changed the title of the PR since 'utilities' was misspelled there.

jsarrao committed:
[kit-plot] Fix error bar calculations and added more comments

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,541 @@
+# Parse the command line arguments. Returns the parsed object if parsing was
+# successful and will terminate the program with an error message otherwise
+def parse_command_line_args():
+    descr = "Generate plots from a Kitsune test suite report. This will overwrite any plot files currently in the save directory."
```
Please format this line to be at most 80 characters wide. This is in keeping with LLVM's conventions which I tend to use.

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,541 @@
+
+# If any benchmarks are added to `Kitsune/Benchmarks`, this list should be
+# updated
+
```
```suggestion
```

tarunprabhu replied:
There are too many empty lines here. There is a balance to be struck between too many empty lines and too few. It's hard to get this right and a lot of it depends on the fussiness of whoever wrote the style conventions for a project. In this case, removing some of these empty lines would be better.
```suggestion
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,541 @@
+BENCHMARKS = ["copy", "euler3d", "raytracer", "saxpy", "srad", "vecadd"]
+
+# All known metrics collected when building/running the kitsune-test-suite
+
```
```suggestion
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,541 @@
+# If any additional metrics are collected, this list should be updated. In most
+# cases, it will also be necessary to add support for metric in
+# parse_and_plot_benchmark_data()
+
```
```suggestion
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,541 @@
+METRICS = ["compile-time", "link-time", "total-runtime", "mean-runtime"]
+
+# All known benchmarks in kitsune-test-suite that have multiple kernels
+
```
```suggestion
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,541 @@
+
+# If any multi-kernel benchmarks are added to `Kitsune/Benchmarks`, this list
+# should be updated
+
```
```suggestion
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,541 @@
+
+# All known metrics collected from kernels in multi-kernel benchmarks when
+# building/running the kitsune-test-suite
+
```
```suggestion
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,541 @@
+# If any additional metrics are collected, this list should be updated. In most
+# cases, it will also be necessary to add support for metric in
+# parse_and_plot_benchmark_data()
+
```
```suggestion
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,541 @@
+
+# All known targets that are suitible to act as baselines for normalizing
+# kitsune-test-suite data.
+
```
```suggestion
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,541 @@
+# If any additional targets are suitable to be baselines, this list should be
+# update. Even if a report has data for more than one baseline target, only
+# one baseline will be selected for plotting
+
```
```suggestion
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,541 @@
+BASELINE_TARGETS = ["kokkos-nvidia", "kokkos-amd"]
+
+# Figure width and height, in inches
+
```
```suggestion
```

tarunprabhu commented on Kitsune/utils/kit-plot:
```diff
@@ -0,0 +1,541 @@
+    parser.add_argument(
+        "--filter-extremes",
+        action="store_true",
+        help="Subtract min and max values from total runtime data. Mean runtimes will also be recalulated from this new total and decreased count",
```
Line width

tarunprabhu reviewed (changes requested):
This looks fine except for some formatting-related nits. Thanks for seeing this through.

jsarrao committed:
[kit-plot] fixed whitespace and line width]

tarunprabhu reviewed (approved):
Thanks a lot for seeing this through. I'll merge this in at some point.
